Bottom: f3cb66112e35dfb3fb7ce2d5156ccb087e498c47
Top:    3923469011f7d3bfe815d63a38b3cb5b3916da5b
Author: Robert Collins <robertc@robertcollins.net>
Date:   2013-02-17 16:37:54 +1300

Refresh of test-control

---

diff --git a/NEWS b/NEWS
index 0b9ebc3..addff88 100644
--- a/NEWS
+++ b/NEWS
@@ -20,6 +20,9 @@ Improvements
   ``StreamResult`` objects (each of which receives all the events).
   (Robert Collins)
 
+* New support class ``TestControl`` which permits cancelling an in-progress
+  run. (Robert Collins)
+
 * New test support class ``testtools.testresult.doubles.StreamResult``, which
   captures all the StreamResult events. (Robert Collins)
 
diff --git a/testtools/__init__.py b/testtools/__init__.py
index 2ea5542..0fee56d 100644
--- a/testtools/__init__.py
+++ b/testtools/__init__.py
@@ -28,6 +28,7 @@ __all__ = [
     'skipUnless',
     'StreamResult',
     'StreamSummary',
+    'TestControl',
     'ThreadsafeForwardingResult',
     'try_import',
     'try_imports',
@@ -74,6 +75,7 @@ else:
         StreamSummary,
         Tagger,
         TestByTestResult,
+        TestControl,
         TestResult,
         TestResultDecorator,
         TextTestResult,
diff --git a/testtools/testresult/__init__.py b/testtools/testresult/__init__.py
index 9aa6b7e..a42439b 100644
--- a/testtools/testresult/__init__.py
+++ b/testtools/testresult/__init__.py
@@ -10,6 +10,7 @@ __all__ = [
     'StreamSummary',
     'Tagger',
     'TestByTestResult',
+    'TestControl',
     'TestResult',
     'TestResultDecorator',
     'TextTestResult',
@@ -24,6 +25,7 @@ from testtools.testresult.real import (
     StreamSummary,
     Tagger,
     TestByTestResult,
+    TestControl,
     TestResult,
     TestResultDecorator,
     TextTestResult,
diff --git a/testtools/testresult/real.py b/testtools/testresult/real.py
index fe49701..c194386 100644
--- a/testtools/testresult/real.py
+++ b/testtools/testresult/real.py
@@ -8,6 +8,7 @@ __all__ = [
     'MultiTestResult',
     'StreamResult',
     'Tagger',
+    'TestControl',
     'TestResult',
     'TestResultDecorator',
     'ThreadsafeForwardingResult',
@@ -522,6 +523,23 @@ class StreamSummary(StreamResult):
             not self.expectedFailures and not self.unexpectedSuccesses)
 
 
+class TestControl(object):
+    """Controls a running test run, allowing it to be interrupted.
+    
+    :attribute shouldStop: If True, tests should no run and should instead
+        return without running. Similarly a TestSuite should abort running
+        and return immediately.
+    """
+
+    def __init__(self):
+        super(TestControl, self).__init__()
+        self.shouldStop = False
+
+    def stop(self):
+        """Indicate that tests should stop running."""
+        self.shouldStop = True
+
+
 class MultiTestResult(TestResult):
     """A test result that dispatches to many test results."""
 
diff --git a/testtools/tests/test_testresult.py b/testtools/tests/test_testresult.py
index 183b184..ae377db 100644
--- a/testtools/tests/test_testresult.py
+++ b/testtools/tests/test_testresult.py
@@ -26,6 +26,7 @@ from testtools import (
     StreamSummary,
     Tagger,
     TestCase,
+    TestControl,
     TestResult,
     TestResultDecorator,
     TestByTestResult,
@@ -785,6 +786,17 @@ testtools.matchers._impl.MismatchError: Differences: [
         self.assertEqual("foo.bar", result.unexpectedSuccesses[0].id())
 
 
+class TestTestControl(TestCase):
+
+    def test_default(self):
+        self.assertEqual(False, TestControl().shouldStop)
+
+    def test_stop(self):
+        control = TestControl()
+        control.stop()
+        self.assertEqual(True, control.shouldStop)
+
+
 class TestTestResult(TestCase):
     """Tests for 'TestResult'."""
